KOGITO-7453: Update workflow error handling guide to describe new ExceptionHandlerPolicy (#51)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc

An link:{spec_doc_url}#error-definition[error definition] in Serverless Workflow is composed of a `name` and `code` parameters. The `name` is a short and natural language description of an error, such as `wrong parameter`. The `code` parameter helps the implementation to identify the error.

> In Kogito, the `code` parameter is mandatory, and the engine uses different strategies to map the provided value to an exception encountered at runtime:

Author: thanks for the idea, I have applied the change

> In Kogito, the `code` parameter is mandatory and the engine uses different strategies to map the provided value to an exception encountered at runtime. The available strategies include _Fully Qualified Class Name_, _Error Message_ and _Status Code_ and are described in more details below.

Author: done

> === Fully Qualified Class Name (FQCN)

Reviewer: ```suggestion
Fully Qualified Class Name (FQCN)::
+
--
```

Author: done

> Define the mapping as the fully qualified Java class name of an link:{java_exception_url}[exception type].

Reviewer: ```suggestion
You can define the mapping as the fully qualified Java class name of an link:{java_exception_url}[exception type].
```

Author: done

> To determine if an exception is handled by an error definition, Kogito first checks whether or not the type of the Java exception is an instance of the exception that is specified in the error definition. Otherwise, Kogito recursively applies the same check to any embedded exception within the original Java one, till the exception chain is exhausted.

Reviewer: ```suggestion
To determine if an exception is handled by an error definition, Serverless Workflow first checks whether or not the type of the Java exception is an instance of the exception that is specified in the error definition. Otherwise, Serverless Workflow recursively applies the same check to any embedded exception within the original Java one, till the exception chain is exhausted.
```

> **Examples**:

Reviewer: ```suggestion
.Examples of runtime exceptions
```

> [options="header"]

Reviewer: ```suggestion
[cols="50%,50%", options="header"]
```

> |=======================

Reviewer: ```suggestion
|===
```

Author: done

|Runtime exception|Error code definition
|`java.lang.RuntimeException`|`java.lang.RuntimeException`
|`java.lang.IllegalStateException`|`java.lang.RuntimeException`
|=======================

> ==== Error message

Reviewer 2: is this intentionally skipping one level of headers? i.e. `=== Error message`

Author: I have changed the header level

> |=======================

Reviewer: ```suggestion
|===
--
```

Author: done

> === Error message

Reviewer: ```suggestion
Error message::
+
--
```

> Define the mapping as a partial match of the error message using link:{java_regex_url}[regex] patterns.

Reviewer: ```suggestion
You can define the mapping as a partial match of an error message using the link:{java_regex_url}[regex] patterns.
```

> Kogito checks whether the message of the Java exception matches the value provided in the error `code`. The validation is performed using a regex matcher, and will be successful if a partial match is found. Note that the validation is performed in a case-sensitive manner. If case-insensitive pattern matching is desired, use the link:{java_regex_pattern_url}[embedded flag expression] `(?i)` within the regular expression.

Reviewer: ```suggestion
Serverless Workflow verifies whether the message of the Java exception matches the value provided in the error code. The validation is performed using a regex matcher. The validation is successful if a partial match is found. Note that the validation is performed in a case-sensitive manner. If case-insensitive pattern matching is desired, use the link:{java_regex_pattern_url}[embedded flag expression] `(?i)` within the regular expression.
```

> **Examples**:

Reviewer: ```suggestion
.Examples of error messages
```

> [options="header"]

Reviewer: ```suggestion
[cols="50%,50%", options="header"]
```

> |=======================

Reviewer: ```suggestion
|===
```

|Error message of runtime exception|Error code definition
|`Unknown error, status code 400`|`Unknown error`
|`Unknown error, status code 400`|`(?i)Error` or `error`
|`Unknown error, status code 400`|`(.*)status code 4[0-9]\{2\}` or `status code 4[0-9]\{2\}`

> |=======================

Reviewer: ```suggestion
|===
--
```

> === Status code

Reviewer: ```suggestion
Status code::
+
--
```

> Define the mapping as a status code returned by the invoked service.

Reviewer: ```suggestion
You can define the mapping as a status code, which is returned by an invoked service.
```

Author: done

> External services could return status codes in the event of a failure. Typical example are HTTP status codes returned by REST or OpenAPI service invocations. In this case, Kogito checks the defined error `code` against the HTTP code from the underlying response object. The error code can be defined either as the HTTP code itself, or as colon separated string where the error code is in the last part of the error code definition.

Reviewer: ```suggestion
The external services return status codes in the event of a failure. For example, HTTP status codes are returned by REST or OpenAPI service invocations. In this case, Serverless Workflow verifies the defined error code against the HTTP code from the underlying response object. 

You can define the error code either as an HTTP code or as a colon-separated string, in which the error definition contains the error code in the last part. 
```

Author: done

> **Examples**:

Reviewer: ```suggestion
.Examples of status code
```

Author: done

> [options="header"]

Reviewer: ```suggestion
[cols="50%,50%", options="header"]
```

Author: done

> |=======================

Reviewer: ```suggestion
|===
```

Author: done

|HTTP status code|Error code definition
|`400`|`400`
|`400`|`HTTP:400`

> |=======================

Reviewer: ```suggestion
|===
--
```
